### ISSUE TITLE
Make Db.SQL enumerator work across async/await continuation

### ISSUE THREAD

**Issue by Razkan** [Starcounter/Home#443]
```
Starcounter.DbException: ScErrIteratorNotOwned (SCERR4035): A cursor operation failed because the cursor wasn't owned by the current thread and transaction.\r\nVersion: 2.3.2.930.\r\nHelp page: https://docs.starcounter.io/v/2.3.1/?q=SCERR4035.\r\n   
at Starcounter.Enumerator.Dispose() in C:\\TeamCity\\BuildAgent\\work\\sc-17406\\Level1\\src\\Starcounter\\Mock\\Enumerator.cs:line 148\r\n   
at Starcounter.Query.Execution.IndexScan.Reset(Row obj) 
in C:\\TeamCity\\BuildAgent\\work\\sc-17406\\Level1\\src\\Starcounter\\Query\\Execution\\Enumerators\\IndexScan.cs:line 778\r\n   
at Starcounter.Query.Execution.ExecutionEnumerator.Dispose() 
in C:\\TeamCity\\BuildAgent\\work\\sc-17406\\Level1\\src\\Starcounter\\Query\\Execution\\Enumerators\\ExecutionEnumerator.cs:line 470\r\n   
at Starcounter.SqlEnumerator`1.Dispose() 
in C:\\TeamCity\\BuildAgent\\work\\sc-17406\\Level1\\src\\Starcounter\\Query\\Execution\\Enumerators\\GenericEnumerator.cs:line 53\r\n   
at FacebookManager.Database.Campaign.<Sync>d__74.MoveNext() 
in C:\\Mopedo\\DRTB\\Facebook Manager\\Database\\Campaign.cs:line 200\r\n--- End of stack trace from previous location where exception was thrown ---\r\n   
at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()\r\n   
at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)\r\n   
at FacebookManager.Protocol.MFC1.Worker.FacebookSyncerWorker.<Run>d__1.MoveNext() 
in C:\\Mopedo\\DRTB\\Facebook Manager\\Protocol\\MFC1\\Worker\\Worker.cs:line 521"
```

Getting this error when a worker is done with what I presume is `Scheduling.RunTask`. I can reporduce this locally.
Everything done within the `Scheduling.RunTask` is committed to the db.

Anything additionally you need?

**Comment by bigwad** [comment]
@Razkan, can you share a code to reproduce the issue?

Most probably you're trying to access an enumerator (which is a result of  `Db.SQL` call) outside of the scope of a transaction where you originally get it. Async/await in foreach loop maybe?

**Comment by Razkan** [comment]
@bigwad, I do not have a running example of the issue, but I can share you the specific source code files where the problem raised from.

Async/await being the problem sounds about right, it was the last thing we changed with the code.

**Comment by miyconst** [comment]
@Razkan please see if you are hitting this limitation - [Short-running transactions - Limistations](https://docs.starcounter.io/topic-guides/transactions/short-running-transactions#limitations).

**Comment by Razkan** [comment]
@miyconst I have removed all `Db.TransactAsync` within the async/await call and still receiving the same error.

@bigwad i've done some further testing and it is as you say as soon as the enumerator is done with the first object the issue arise. The issue is reproducable in production code, but not when I try to create an seperate example. However the seperate example produces an error of itself, probably in the similar manners(?)

```
using System;
using System.Collections.Generic;
using System.Net.Http;
using System.Threading.Tasks;
using Starcounter;
using Starcounter.Logging;

namespace DbTest2
{
    class Program
    {
        static void Main()
        {
            new Program().Run();
        }

        public async void Run()
        {
            Db.Transact(() =>
            {
                var a = new A();
                new B {A = a};
                new B {A = a};
            });
            try
            {
                await Db.SQL<A>($"SELECT a FROM {typeof(A)} a").First().Sync();
            }
            catch (Exception e)
            {
            }
        }
    }

    [Database]
    public class Base
    {
        protected static ISyncer Syncer { get; } = new Syncer();

        public bool Done { get; set; }
    }

    public class A : Base
    {
        internal async Task Sync()
        {
            await Syncer.Sync(this);
            foreach (var b in B)
                await b.Sync();
        }

        public IEnumerable<B> B => Db.SQL<B>("SELECT t from DbTest2.B t where t.A =?", this);
    }

    public class B : Base
    {
        public A A { get; set; }

        internal async Task Sync()
        {
            await Syncer.Sync(this);
        }
    }

    internal class Syncer : ISyncer
    {
        Task ISyncer.Sync<T>(T entity) => Scheduling.RunTask(() =>
        {
            var t = new HttpClient().GetAsync("https://www.google.com/").Result;
            Db.Transact(() => entity.Done = true);
        });
    }

    public interface ISyncer
    {
        Task Sync<T>(T entity) where T : Base;
    }
}
```

which produces

```
Starcounter.DbException: ScErrIteratorClosed (SCERR4139): A interator operation failed because the iterator was closed.
Version: 2.3.2.930.
Help page: https://docs.starcounter.io/v/2.3.1/?q=SCERR4139.
   at Starcounter.FilterEnumerator.MoveNext() in C:\TeamCity\BuildAgent\work\sc-17406\Level1\src\Starcounter\Mock\Enumerator.cs:line 542
   at Starcounter.Query.Execution.FullTableScan.MoveNext() in C:\TeamCity\BuildAgent\work\sc-17406\Level1\src\Starcounter\Query\Execution\Enumerators\FullTableScan.cs:line 472
   at DbTest2.A.<Sync>d__0.MoveNext() in C:\Users\Sebastian\source\
epos\BuggTest\DbTest2\Program.cs:line 52
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.GetResult()
   at DbTest2.Program.<Run>d__1.MoveNext() in C:\Users\Sebastian\source\
epos\BuggTest\DbTest2\Program.cs:line 28
```

The code i've tested is extremely similar to the code I have in production.

**Comment by bigwad** [comment]
@Razkan, `Db.TransactAsync` is unrelated to your issue. The root cause is that starcounter transaction doesn't flow along with await continuation. I'll come up with a workaround soon. Stay tuned.

**Comment by bigwad** [comment]
Unfortunately, I have to conclude that Db.SQL results enumerating is not async ready. No matter if one uses `foreach` or manual loop with `Enumerator.MoveNext()`. Each `Enumerator.MoveNext()`  should be called on exactly the same thread where enumerator was originally acquired. What I can recommend is to retrieve all objects to be processed in a collection and then process this collection asynchronously.
Please observe changes in `Main`, `A::Sync` and `Syncer::Sync`

```csharp
using System;
using System.Collections.Generic;
using System.Linq;
using System.Net.Http;
using System.Threading.Tasks;
using Starcounter;

namespace DbTest2
{
    class Program
    {
        static long NumberOfItemsNotDone()
        {
            return Db.SQL<long>("select count(b) from B b where b.Done=?", false).First();
        }
        static void CheckPreCondition()
        {
            System.Diagnostics.Debug.Assert(NumberOfItemsNotDone() != 0);
        }

        static void CheckPostCondition()
        {
            System.Diagnostics.Debug.Assert(NumberOfItemsNotDone() == 0);
        }

        static void Prepare()
        {
            Db.Transact(() =>
            {
                foreach (var a in Db.SQL<A>("select a from A a"))
                    a.Delete();
                foreach ( var b in Db.SQL<B>("select b from B b"))
                    b.Delete();
            });
        }

        static void Main()
        {
            Prepare();

            new Program().Run().Wait();
            CheckPostCondition();

            System.Environment.Exit(0);
        }

        public async Task Run()
        {
            Db.Transact(() =>
            {
                var a = new A();
                new B { A = a };
                new B { A = a };
            });

            CheckPreCondition();

            await Db.SQL<A>($"SELECT a FROM {typeof(A)} a").First().Sync();
        }
    }

    [Database]
    public class Base
    {
        protected static ISyncer Syncer { get; } = new Syncer();

        public bool Done { get; set; }
    }

    public class A : Base
    {
        internal async Task Sync()
        {
            await Syncer.Sync(this);

            await Scheduling.RunTask(async () =>
            {
                foreach (var b in B.ToList())
                {
                    await Scheduling.RunTask(async () => await b.Sync());
                }
            });
        }

        public IEnumerable<B> B => Db.SQL<B>("SELECT t from DbTest2.B t where t.A =?", this);
    }

    public class B : Base
    {
        public A A { get; set; }

        internal async Task Sync()
        {
            await Syncer.Sync(this);
        }
    }

    internal class Syncer : ISyncer
    {
        async Task ISyncer.Sync<T>(T entity)
        {
            await new HttpClient().GetAsync("https://www.google.com/");

            await Scheduling.RunTask(async () => await Db.TransactAsync(() => entity.Done = true) );
        }
    }

    public interface ISyncer
    {
        Task Sync<T>(T entity) where T : Base;
    }
}
```

**Comment by Razkan** [comment]
@miyconst @bigwad thanks, for the recommendation!

We got it to work by doing the following change:
```cs
foreach (var b in B)
	await b.Sync();
```
To:
```cs
foreach (var b in B.ToArray())
	await b.Sync();
```
Forcing the enumerator to get all the entities in the same context before using them.

**Comment by bigwad** [comment]
Just please note, that call to `b.Sync()` has to be wrapped in `Scheduling.RunTask` if you refer any b's properties in the Sync. In your case you don't, but just to be on the safe side, I'd call it using the wrapper: 
```csharp
await Scheduling.RunTask(async () => await b.Sync());
```

**Comment by miyconst** [comment]
Related - [async/await triggers ScErrThreadNotAttached](https://github.com/Starcounter/level1/issues/2796).

**Comment by bigwad** [comment]
>Do we need a separate transaction inside the RunTask() delegate (given that we need to change values of b (which I think is the case in the production code))?

Yes, and I think you do in `Syncer.Sync`:
```scharp
            Db.Transact(() => entity.Done = true);
```

>This should work too, right?
foreach (var id in B.Select(b => b.GetObjectNo()).ToArray())
    await Db.FromId<T>(id).Sync();

Yes, it will also work, but I suspect more overhead because of two instantiations of each B. Other than that there is no practical difference.

>If so, I think it is the preferred way to deal with this without leaving the transaction scope. Or maybe we do anyways since we call await? Please enlighten us 😃

I don't see how this code changes transaction/enumerator/thread interrelations. The code after `await` point is still highly likely to run on a different thread. If you want it to share a transaction with the code that enumerates B's - special measures have to be taken (though only long-running transaction are possible to share in 2.3).

**Comment by miyconst** [comment]
Closing as won't fix in Starcounter 2.4 and not applicable in Starcounter 3.0.
